### PR TITLE
OAK-9790 Implement parallel indexing for speeding up oak run indexing command

### PR DESCRIPTION
**Pull request by thomasmueller**
(This is the PR https://github.com/apache/jackrabbit-oak/pull/587 but merged into a branch of Oak)

Since indexing was single threads, which is slow for large repository. In order to improve the indexing performance we need to implement parallel indexing.

The work is cover for both lucene and elastic indexing. In order to support parallel indexing, it need to split the big flat file store file ahead, which add a big overhead, but make parallel index possible and much faster.

Another change together is support the LZ4 compression since which is much faster compare to gzip.

### CODE REVIEW

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java:63]: BULK_PROCESSOR_CONCURRENCY should be configurable and not set to 4 as default as this would impact async indexing as well -- see https://github.com/apache/jackrabbit-oak/pull/587#pullrequestreview-1022678685

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticConnection.java
Comment (thomasmueller) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/ElasticConnection.java:124]: Why do we need to change the socket timeout?

File: oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java
Comment (nit0906) [inline oak-search-elastic/src/main/java/org/apache/jackrabbit/oak/plugins/index/elastic/index/ElasticBulkProcessorHandler.java:63]: This PR should not be changing ElasticBulkProcessorHandler like this, since this would directly impact async indexing as well.

If we need to have a BP concurrency here, I suggest we make this configurable and set it via the oak-run OutOfIndexing flow (which this PR targets) and keep the default concurrency to 1 (i.e no concurrency).

File: oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/progress/IndexingProgressReporter.java
Comment (joerghoh) [inline oak-core/src/main/java/org/apache/jackrabbit/oak/plugins/index/progress/IndexingProgressReporter.java]:
```suggestion
                    log.info("{} => Indexed {} nodes in {} ...", indexPath, count, watch);
```

File: oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java
Comment (joerghoh) [inline oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java]: It's safe to use Java 8's ```Objects.requireNonNull()``` here. We should avoid adding even more references to guava.

File: oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java
Comment (joerghoh) [inline oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java:197]: is this a comment (that means helpful to understand the code) or just commented code, which has been created during development but which is not used here? In the later case, please remove it.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java]: it would make sense to merge these 2 INFO messages into a single one.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java]: this should also live in a ```finally```  clause, otherwise it might left open.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java]: please merge these 2 INFO messages into a single statement.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java]: this loop is a bit inconsistent to the previous check (```nodeTypeNameStack.contains("")```), because it does not iterate over the last element in the stack, but the ```contains()``` does); is this intendend and valid?

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreUtils.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreUtils.java]:
```suggestion
     * This function by default uses gzip as compression algorithm for backward compatibility.
```

The term ```gzip``` is  more widely used than ```GNU zip```

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreUtils.java
Comment (joerghoh) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileStoreUtils.java]:
```suggestion
     * This function by default uses gzip as compression algorithm for backward compatibility.
```

**Review by joerghoh — CHANGES_REQUESTED** [review]

File: oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java
Comment (thomasmueller) [reply oak-lucene/src/main/java/org/apache/jackrabbit/oak/plugins/index/lucene/writer/DefaultIndexWriter.java:197]: Jun didn't change the old code here (from Chetan / Vikas / Tommaso probably), he just synchronized the method. So I think we should keep it as is. Unless if you can find out if the comment is better kept or removed... I can't, looking at the code.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (thomasmueller) [reply oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java]: The code itself is correct (at least this line is correct), but it is not properly documented.

File: oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java
Comment (thomasmueller) [inline oak-run-commons/src/main/java/org/apache/jackrabbit/oak/index/indexer/document/flatfile/FlatFileSplitter.java:196]: (I added this TODO on purpose: we can't merge the PR currently, because this still needs to be implemented.)

The code only deals with primary types (e.g. acme:Asset), but doesn't deal with mixing types.
We need to work on this, before we can merge the patch.
Using a stack alone isn't sufficient: we need a stack of sets :-/

**Review by thomasmueller — COMMENTED** [review]
The PR is NOT ready to be merged. But the review can continue.

**Comment by amit-jain** [comment]
Superseded by PR https://github.com/apache/jackrabbit-oak/pull/715

**Pull request closed without merging**
